### ISSUE TITLE
Best practices for change detection

### ISSUE THREAD

**Issue by GillesVercammen** [ngx-formly/ngx-formly#1641]
[ ] bug report => search github for a similar issue or PR before submitting
[ ] feature request
[x] support request

Hello, 
We are using formly in a large enterprise environment at the moment.
While using formly we stumbled upon some problems we didn't anticipate and we are looking to find out best practices for some implementations.

Some problems: 

1) The property:  ExpressionProperties is being used extensivly in our project. 
We use it to handle all kind of changes and even select values from the ngrx store.
We started noticing that this property gets called  a lot, when loading the page initially. It gets called 25+ times.
So we were wondering if this is the correct place to track changes on the field?

2) we tried using hooks (onInit, onChanges,  etc), but onChanges was not being called when values in the fields changes or select drop down was changed. So this was not useable for us.

3) we have yet to find a good way to format value that is coming from the api. 
Eg: we receive from the api:
months: 15

And in the field we want to show: 1 year 3 months.

Currently we do this in the expressionproperties again, but like i said. that function to format the months is being called 25 times on load, plus multiple times after.

So in short, we are looking from some best practices using formly to change fields depending on other fields or apply some logic to the data we receive from the back to present to the view.

Tnx in advance.

**Comment by aitboudad** [comment]
> The property: ExpressionProperties is being used extensivly in our project.
We use it to handle all kind of changes and even select values from the ngrx store.
We started noticing that this property gets called a lot, when loading the page initially. It gets called 25+ times.

that's a known issue, the fix is to add an opt-in to run expression on model changes only if you want to check if that would improve the perf comment the following line https://github.com/ngx-formly/ngx-formly/blob/v5/src/core/src/lib/components/formly.form.ts#L68

> we tried using hooks (onInit, onChanges, etc), but onChanges was not being called when values in the fields changes or select drop down was changed. So this was not useable for us.

for hooks, I would suggest avoiding `onChanges` and only rely `onInit` and `onDestroy` for others I'm thinking to deprecate them.

> we have yet to find a good way to format value that is coming from the api.
Eg: we receive from the api:
months: 15

eaither wait until we provide that option or use an optimized way which rely on the formatted value (on model change only) see the following example https://stackblitz.com/edit/angular-p61kht

**Comment by aitboudad** [comment]
a new option is introduced in `v5.3` to run expression on model changes only, check my comment 
https://github.com/ngx-formly/ngx-formly/issues/1620#issuecomment-514639091 for more details
